### ISSUE TITLE
Better Convergence Monitoring — Dynamic Plots and What to Plot?

### ISSUE THREAD

**Issue by NicolasGensollen** [aramis-lab/leaspy#68]
In GitLab by @little_nemo on Apr 28, 2022, 15:11

*This is a quick draft to consign a point the we discussed orally together with @PEPoulet @JulietteOrtholand & @emaheux.*

## Problems with convergence monitoring

Currently, we monitor the convergence of Leaspy models using PDFs files, that are generated and updated during the convergence. Those files are generated through a [`Plotter`](https://gitlab.com/icm-institute/aramislab/leaspy/-/blob/master/leaspy/io/logs/visualization/plotter.py#L19) object that is instantiated in the [`FitOutputManager`](https://gitlab.com/icm-institute/aramislab/leaspy/-/blob/master/leaspy/io/logs/fit_output_manager.py) class. This class also stores the last few sampling iterations over the Markov Chain in CSV files which are not super easy to manipulate.

This way of handling the logs comes with a few issues:
- PDF files are not always easy to visualize (see the issue we faced during the Winter School to visualize PDFs in Colab)
- when the «re-plotting» frequency is high, a significant amount of the calibration computing time is spent just generating the plots
- those plots are not dynamic at all (we cannot change from linear scale to log scale on demand, zoom on a particular area, hide or identify features, etc)
- some of the logging parameters (such as the length of the Markov Chain that we store -- currently the 100 last iterations) are hard-coded somewhere (note that this point might be partially fixed in the ordinal branch at some point)
- we currently do not directly plot some relevant metrics (likelihood, regularization, ...)

Those issue (and some other that I might be forgetting...) make it difficult to:
- clearly identify whether the model has converged correctly or not
- monitor the convergence during calibration
- adjust some of the hyper-parameters of the model (we have no plot linking the convergence with the temperature, with the memory decaying term, etc) or even way to measure whether we used enough iterations
- dynamically study some aspects of the convergence (plot the distribution of parameters over Markov Chain sampling, compare specific features, etc)

Some of those problems have already been identified in or are linked with previous issues, such as https://gitlab.com/icm-institute/aramislab/leaspy/-/issues/51#note_711708779, https://gitlab.com/icm-institute/aramislab/leaspy/-/issues/27

## Possible Solutions

Here is where discussion is needed.

One point that we identified is that it might be easy to plug some of the points that we want to monitor during convergence into a Tensorboard like interface. I quickly looked at whether it was possible, from what I gathered it might be even easier to set-up a Tensorboard-compatible logger than go through the current `Plotter` object, see [here for instance](https://pytorch.org/tutorials/recipes/recipes/tensorboard_with_pytorch.html).
This would discharge our codebase from handling the plotting interface, since we could rely on the already implemented functionalities of tensorboard (such as histogram visualization, etc). 

A cons would be that it might add an extra-dependency, which might cause some issues when trying to use Leaspy on platforms where we do not have the full control over the environment. Maybe we could make it optional on this regard, and keep the optional CSV based output in those cases.

There might be other similar interfaces available (and more general maybe, if we also want to have model and patient trajectories plotted), so it might be worth exploring a bit more.

## Future work

I don't really have time for this at the moment, this issue is for future discussions whenever someone has some time to allocate to this. I think that what we should do is:
- List what we would precisely like to be monitored during convergence (likelihood, regularization, distribution of samples over iterations, maybe include new metrics such as the one mentionned by Étienne here https://gitlab.com/icm-institute/aramislab/leaspy/-/issues/63, etc)
- Choose whether to stick with the current Plotter matplotlib based-framework, or use another implementation
- Implement it....... :slight_smile:

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 17, 2022, 09:41

I found that in the gibbs_sampler file there were two hard coded constants:

    # Cf. note on `scale` parameter above (heuristic values)
    STD_SCALE_FACTOR_POP = .01
    STD_SCALE_FACTOR_IND = .5

I have done some tests and I am wondering to what extend these constants have been optimized for every pathologies on every datasets...

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on May 17, 2022, 09:56

Those are only starting scales for the std-dev of samplers, which are then adapted during all the fit to achieve a rolling mean acceptation rate inside the `mean_acceptation_rate_target_bounds` interval ([20%, 40%] by default). So I think you should really not bother about those constants, it may only speed-up the very first runs.

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 17, 2022, 10:00

Ok, what about these values then, I have found that `Lastly, the acceptance rate depends on the problem but typically for 1-d problems, the acceptance rate should be around 44% (around 23% for more than 5 parameters). ` [here](https://jellis18.github.io/post/2018-01-02-mcmc-part1/), shouldn't it be more between 20 and 50% ?

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 17, 2022, 10:06

I have been playing with this parameters in an heuristic way and it change a bit the results after 20000 iterations...

[convergence_ok.pdf](https://my-gitlab-bucket.s3.amazonaws.com/938594763/f98b51b5f2ac3aeb7c6838146e0014fce26dfad29a242351d5a58decf3a6bcbd/convergence_ok.pdf)
[convergence_bad.pdf](https://my-gitlab-bucket.s3.amazonaws.com/938594763/e8c48e67271fa3174781aa2e718db548b1a82cf453a576716329fec8529091d6/convergence_bad.pdf)

I am looking for a more theoretical explanation

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on May 17, 2022, 14:11

It would be interesting to see whether those heuristic changes are in the “same range” as changes coming from a different seed being used for calibration. I imagine that changing those starting scales changes a bit what happens at least at the beginning of the calibration, hence changing how randomness propagates during execution.

> I have been playing with this parameters

Are you talking about `mean_acceptation_rate_target_bounds` or `STD_SCALE_FACTOR_POP` for the experiment that you present?

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 17, 2022, 15:20

I have played with both, I rerun 2 fits with different seed, I'll come back when they'll finish

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 23, 2022, 13:53

[convergence_1.pdf](https://my-gitlab-bucket.s3.amazonaws.com/938594763/74f8773d9c085cab1545d5d2cd0b2a079ead698f17e8f4fb3e428d1f3e9bab12/convergence_1.pdf), I have only changed the seed... results are quite different...

**Comment by NicolasGensollen** [comment]
In GitLab by @little_nemo on May 23, 2022, 14:41

Indeed this run looks surprisingly different from the one in [`convergence_ok.pdf`](https://gitlab.com/icm-institute/aramislab/leaspy/uploads/f8b686a0de191291c431f5dc08587fd9/convergence_ok.pdf) (especially for the noise term which is usually quite stable across different runs...). 
I guess that if the effect of the seed is as significant, it can be enough to explain the discrepancy that you observed when tweaking the hyper-parameters without the need of too much theory.

But, other than that, I am not sure where to really go from here, except that agreeing that there is something going on that we should adress. :|

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 23, 2022, 14:52

I am going to have an empiric look to xi_std and tau_std initialization, see what it could change to modify them as they seems hard coded

**Comment by NicolasGensollen** [comment]
In GitLab by @JulietteOrtholand on May 30, 2022, 13:34

I created a new issue as my issue is more on the convergence itself than in how it is tracked see #70

**Comment by NicolasGensollen** [comment]
In GitLab by @emaheux on Jun 1, 2022, 11:48

> [convergence_1.pdf](https://gitlab.com/icm-institute/aramislab/leaspy/uploads/e45ac9ef8653422795e42fe9a0b196a2/convergence_1.pdf), I have only changed the seed... results are quite different...

You also had 1 feature more in this calibration compared to the first two no?
